### ISSUE TITLE
Proposal: optional argument checking

### ISSUE THREAD

**Issue by matbesancon** [JuliaLang/julia#34704]
The `@inbounds` / `@boundscheck` pair is a special case of letting callers opting in and out of argument checking, this helps making array code fast, while keeping safety on by default.

This proposal is about generalizing this to general argument checking, with the same pattern:
- `@argcheck blk` signaling that a given block is argument checking, which may throw errors or return from the function with an error value, etc
- `@skipargcheck expr` eliminating the above blocks in the following expression.

My intuition is that such construct requires the same kind of compiler support to work in most Julia code bases. A dispatch-based solution would be doable but a bit of a hassle.

**Comment by KristofferC** [comment]
`@inbounds`, `@propagate_inbounds` and `@boundscheck` are pretty weird  in how they work. It is kinda strange how an annotation somewhere can cause a different behaviour somewhere else. Also `@inbounds` (as all performance annotations) has also been awkward to use in generic code 

I personally think that spreading this type of non-standard evaluation to more parts of the language would be a mistake. Seems easier to add a keyword argument `nocheck=true` or something similar.

**Comment by bramtayl** [comment]
```julia
Seems easier to add a keyword argument nocheck=true or something similar
```

I agree that seems easier; maybe we don't even need the `@inbounds` macro at all? Constant propagation through inlined functions should be able to do all the work.

**Comment by matbesancon** [comment]
It is indeed easier to use a keyword argument, but not ideal:
1. the top-caller needs to rely on the fact that all functions call propagate the keyword correctly.
2. `Constant propagation through inlined functions` this requires users to hope that constant propagation kicked in. While it is reasonable to expect it in some cases, relying on a compiler optimization for the semantic of the program feels weird (I am not a PL person so I won't go deeper on that and let folks voice in)

**Comment by simeonschaub** [comment]
The good news is, that, using IRTools, this can all be done in a package as well. I have had a go at it in https://github.com/simeonschaub/OptionalArgChecks.jl/, if you want to check that out.
About constant propagation in keyword arguments: The last time I looked into that, it was very finicky to get it to work reliably. Things might have gotten better in newer versions, but I tend to agree with @matbesancon that relying on these optimizations in performance critical code will probably be quite fragile, especially if the compiler goes through multiple levels of inlining.

**Comment by KristofferC** [comment]
> Constant propagation through inlined functions this requires users to hope that constant propagation kicked in. While it is reasonable to expect it in some cases, relying on a compiler optimization for the semantic of the program feels weird

Agreed which is why just a keyword argument is so much easier / clear. `@inbounds` is a bit special since the whole point is to remove a branch to allow for SIMD. This is not really the case for other checks where a predicted branch is free.

**Comment by matbesancon** [comment]
> Agreed which is why just a keyword argument is so much easier / clear. @inbounds is a bit special since the whole point is to remove a branch to allow for SIMD

I don't understand why it is special compared to other checks. Removing the checking and handling branch also results in branch-free SIMD-friendly code doesn't it?

**Comment by bramtayl** [comment]
Can someone clarify for me whether or not constant propagation would be able to eliminate the branch? If so, maybe we should make constant propagation through inlined functions a semantic guarantee. If not, why not?
